### ISSUE TITLE
[NVPTX] Assertion "Bad value decomposition" hit while compiling tests

### ISSUE THREAD

**Issue by jhuber6** [llvm/llvm-project#83179]
The NVPTX backend currently crashes when trying to compile this test https://github.com/llvm/llvm-project/blob/main/libc/test/src/__support/uint_test.cpp. I have reduced it down to the following LLVM-IR, or in https://godbolt.org/z/TEbxMz5q1.
```llvm
target triple = "nvptx64-nvidia-cuda"

%"struct.BigInt" = type { %"struct.arr" }
%"struct.arr" = type { [2 x i128] }

define void @foo() {
entry:
  %call6 = call %"struct.BigInt" null(ptr null, i64 0)
  ret void
}
```

This currently crashes in the backend, while other targets seem to handle it fine without crashing.

**Comment by Artem-B** [comment]
We're tripped by 'special-ness' of i128 handling. PTX does no have i128, so we have to pass i128 around as a pair of i64.
In this case, we've tripped an assertion that number of elements on both sides must match, and for i128 it does not.

Support for i128 in general is very limited. It's mostly bolted on with a handful of special cases. We can probably relax the assertion here, and things may just work, but if your code depends on i128, you may expect surprises in other places.

**Comment by jhuber6** [comment]
> We're tripped by 'special-ness' of i128 handling. PTX does no have i128, so we have to pass i128 around as a pair of i64. In this case, we've tripped an assertion that number of elements on both sides must match, and for i128 it does not.
> 
> Support for i128 in general is very limited. It's mostly bolted on with a handful of special cases. We can probably relax the assertion here, and things may just work, but if your code depends on i128, you may expect surprises in other places.

I see, I believe that handling is pretty standard for various weird things, so hopefully it's well enough understood by the rest of LLVM. When you say 'both sides' are you referring to a caller and callee? Or is this a host / device thing. If it's the latter, this is from the `libc` test suite so there's no concept of a host.

**Comment by Artem-B** [comment]
> When you say 'both sides' are you referring to a caller and callee?

In this case, LLVM's idea of a type, and the types of the actual chunks we use to load/store the value of that type. They do not always match. 

IIRC, another issue with i128 is that some operations rely on libcalls, which makes them hard to use in the 'classic' CUDA compilation. It will probably be more useful for the stand-alone compilation mode, where you do have ability to link with external libraries.

**Comment by jhuber6** [comment]
> In this case, LLVM's idea of a type, and the types of the actual chunks we use to load/store the value of that type. They do not always match.

I see, thanks.

> IIRC, another issue with i128 is that some operations rely on libcalls, which makes them hard to use in the 'classic' CUDA compilation. It will probably be more useful for the stand-alone compilation mode, where you do have ability to link with external libraries.

I've been floating around the idea of a GPU `compiler-rt` for awhile. I think now that I have the infrastructure set up to simply cross-compile code it's more feasible. We could then just handle this stuff using the existing `-mlink-builtin-bitcode` magic if needed for "classic" CUDA. I finally got around to making a `libc.bc` that can be passed that way for non-linking CUDA.
